### PR TITLE
Deprecate the `FxHashMap()` and `FxHashSet()` constructor function hack

### PR DESCRIPTION
**Pull request by oli-obk**

### CODE REVIEW

**Comment by oli-obk** [comment]
Also, imports won't get fewer, as most of the time you need the type name for the struct fields

**Comment by RalfJung** [comment]
> I'd personally use Default::default() instead of SpecificMapOrSetType::default() where possible, types change from time to time and with Default::default() you don't have to adjust initializers.

Uh, I'd prefer if you wouldn't do that change. That's throwing away useful type information, which can already be hard to get by with all the inference we are doing.

**Comment by RalfJung** [comment]
So, is this fully replacing https://github.com/rust-lang/rust/pull/52591? Looks to be just a partial successor?

**Comment by oli-obk** [comment]
> Looks to be just a partial successor?

Yes that's on purpose. Otherwise I'll never stay ahead of the bitrot curve

**Comment by oli-obk** [comment]
So... how do I continue? :laughing: 

@petrochenkov are you fine with not changing everything to `Default::default`? (well, everything that can be changed, which is around 3/4th I guess)

**Comment by petrochenkov** [comment]
@oli-obk 
Can you do it in field initializers?
That's where it doesn't mess with type inference and is not harmful for readability.

**Comment by RalfJung** [comment]
> is not harmful for readability.

I still think this does more harm then good -- now I have to scroll to the struct definition to figure out the types. That can be arbitrarily far away.

But I won't fight over this.

**Comment by nikomatsakis** [comment]
FWIW, in my own projects, the rule I follow is usually:

- When implementing `Default` (or a `new` function) I use `Default::default()`
    - this is partially because it lets me write simpler Rust macros, emacs macros, or otherwise move quickly through this boilerplate stuff
- Otherwise, I prefer `Type::default()` for the reasons @RalfJung and @oli-obk gave -- I find it more readable, and I don't get surprises when I tweak the code and no longer have sufficient type hints

File: src/bootstrap/cache.rs [hunk 1]
```diff
@@ -169,12 +169,20 @@ impl Ord for Interned<String> {
     }
 }
 
-#[derive(Default)]
 struct TyIntern<T: Hash + Clone + Eq> {
     items: Vec<T>,
     set: HashMap<T, Interned<T>>,
 }
 
+impl<T: Hash + Clone + Eq> Default for TyIntern<T> {
```
Comment (nikomatsakis) [inline src/bootstrap/cache.rs]: Like this, I guess. =)

**Review by nikomatsakis — APPROVED** [review]

**Comment by sanmai-NL** [comment]
Ad @nikomatsakis in https://github.com/rust-lang/rust/pull/55114#issuecomment-430245366:
These kinds of idiom conventions are ephemeral if they’re not written down in https://github.com/rust-lang-nursery/api-guidelines. Discussing it in pull requests won’t capture the knowledge very well and discoverably.

**Comment by nikomatsakis** [comment]
@sanmai-NL true. But that particular comment was just giving my preferences. In any case, I think such docs might be appropriate for rustc-guide -- feels like it doesn't belong in the API guidelines (particularly since these decisions don't affect the API very much).

(That said, I do think there is an API decision here, in that I've come to favor implementing `Default` over implementing a zero-argument `new` fn.)

File: src/bootstrap/cache.rs [hunk 2]
```diff
@@ -169,19 +169,21 @@ impl Ord for Interned<String> {
...
 
-struct TyIntern<T> {
+struct TyIntern<T: Hash + Clone + Eq> {
     items: Vec<T>,
     set: HashMap<T, Interned<T>>,
 }
 
-impl<T: Hash + Clone + Eq> TyIntern<T> {
-    fn new() -> TyIntern<T> {
+impl<T: Hash + Clone + Eq> Default for TyIntern<T> {
```
Comment (bjorn3) [inline src/bootstrap/cache.rs]: Can this be derived?

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: src/bootstrap/cache.rs (hunk 2 above)
Comment (RalfJung) [reply src/bootstrap/cache.rs]: No; the derive `impl` would have an additional `where T: Default`.

A comment saying so might be a good idea though.

Comment (bjorn3) [reply src/bootstrap/cache.rs]: Get it

**Comment by oli-obk** [comment]
@bors r=nikomatsakis p=1

bitrot priority

**Comment by RalfJung** [comment]
Looks like it timed out after 3h.

@bors retry

**Pull request merged**